SQL injection detection

CaspianA1 opened the pull request:
Mostly done, but not quite; I just made a PR to run e2e tests and get feedback on my progress so far.

Here's a summary of my changes:
1. Improved global variable mutation detection (can now detect weird variable swaps, like `z = [y, y = z][0];`, and modifications separated by a comment, like `x = 23 + x, y.a = 24 + x;`). This covers all assignment expression types, I believe, except for spread assignments, which I will need to implement at some point.
2. Minor function name changes, README changes, stuff like that
3. Lots of SQL injection detection code
4. Add lots of test cases for this
5. Some dependency tweaks, to make the chances of broken installs of this package yet smaller

Here's how I determine probable SQL injections:
- If I detect a callsite where raw SQL queries can happen (e.g. `ctxt.client.raw(myQuery)`, where the ORM client is `Knex`), figure out if the value is reducible to a string literal, or concatenations of string literals.
- This means that the parameter can be a string literal, a variable that is only ever assigned to string literals, a concatenation of string literals, a variable that's assigned to concatenations of string literals with other such variables, etc...
- Template expressions (e.g. ``` `${foo}, ${bar}, ${baz}` ```) are also supported, as long as the values used to build it (`foo`, `bar`, and `baz` here) are also reducible to string literals. This does not include tagged template expressions, since those involve a function call, and my variable definition analysis stays within the scope of a single `Transaction` function.
- Using any of a `Transaction` function's parameters for a raw SQL query is not allowed either, since you don't know who will call that transaction except in certain cases.
- If the passed-in variable to the query is not local, I examine the whole scope of the program for any non-literal-reducible assignments. If I find any, then I raise an error.
- If some variable passed into a callsite for raw SQL queries does not meet the aforementioned requirements, I flag it as a SQL injection, because that then means that you most likely took some datatype and cast it to a string to build your query, which is risky.

Here's what's left:
- [x] Fixing some versioning hassle
- [x] Add support for more callsites (beyond just `Knex`'s `raw` function). Should be relatively easy.
- [x] Proper line reporting for errors (provide the line where the potential injection happened, not the line of the raw SQL call). Medium difficulty (not particularly hard, it'd just take an hour or two to implement).
- [x] Dealing with type aliasing

And things to think about after that:
- [ ] Maybe split everything into more files (but that might be another PR)
- [x] Version bump

CaspianA1 committed:
A couple of things:
- Change some function and type names
- Prepare for detecting SQL injection

CaspianA1 committed:
Make the README a bit more correct

CaspianA1 committed:
Fix 'checkSqlInjectionDecorators'. Set up a test for SQL injection detection. Flesh another function out slightly more.

CaspianA1 committed:
Flesh out more of the SQL injection detection logic

CaspianA1 committed:
Make a comment more correct

CaspianA1 committed:
Word some tests better. Start on the SQL injection code (just two cases, and for Knex).

CaspianA1 committed:
Change some return styles (it was annoying to read otherwise)

CaspianA1 committed:
Move some core util functions

CaspianA1 committed:
Now ignoring dbos-demo-apps from local e2e tests in the gitignore and with vitest

CaspianA1 committed:
Change an error message

CaspianA1 committed:
make it so that ts-morph typenames can never be undefined (this was true before, I just never noticed this earlier)

CaspianA1 committed:
Now throwing an error when there's no ESLint node that's needed

CaspianA1 committed:
Expand the test suite. Make the injection detection logic much more robust (although there are still some more cases to handle). Replace `getLocal` with `isLocal` again.

CaspianA1 committed:
Big commit. Main changes:
- Expand the test cases
- Now inferring when assignment is happening correctly (it was very half-assed before, and I had some differing logic on how to handle it - both incomplete ways! - in the global variable detection code and the SQL injection detection code)
- General improvements: supporting all assignment operator types, global variable detection can now handle one of the weird variable swap cases, and the stack frames are now sets again, instead of maps.

CaspianA1 committed:
Remove some whitespace

CaspianA1 committed:
Add a plan on how to detect SQL injection

CaspianA1 committed:
The core injection detection is now done! Really proud of how the logic worked out, it's quite elegant IMO.
- I now just need to clean up a bit of the code, support more callsites, disallow format strings, and test some other binary operators on strings too.
- Also replace all throws with a `panic` function that I made (following DRY!).

CaspianA1 committed:
Remove a part of some function names in the tests

CaspianA1 committed:
Remove some code repetition

CaspianA1 committed:
Some further test utility renaming

CaspianA1 committed:
Split up some shared types and values. Make some more shared constants global (for easy tweaking in the future). Remove a comment, and add a TODO.

CaspianA1 committed:
Rename some types and variables

CaspianA1 committed:
More code cleanup

CaspianA1 committed:
Document the justification for how the literal-reducible string stuff detect SQL injection in a more clear way. More type renaming.

CaspianA1 committed:
Simplify the LR state that I'm caching

CaspianA1 committed:
Clean up 'checkCallForInjection' substantially

CaspianA1 committed:
Make the tests for SQL injection more meaningful in terms of why each one is there

CaspianA1 committed:
Half-remove a TODO

CaspianA1 committed:
Make getRValuesAssignedToIdentifier a full generator (no intermediate allocation is happening anymore)

CaspianA1 committed:
Make 'analyzeFunction' more streamlined (it now just goes through a decorator set error checker mapping to evaluate different determinism checkers)

CaspianA1 committed:
Find some failure cases, and make a note about console.log

CaspianA1 committed:
Now making using function parameters as raw sql query parameters illegal. Also rename some functions to make them clearer.

CaspianA1 committed:
Whittle away some TODOs

CaspianA1 committed:
Now supporting template expressions when the data within them is also literal-reducible

CaspianA1 committed:
Some minor reformatting

CaspianA1 committed:
Fixed some unresolved symbol errors in the tests. Also made a wrapper around symbol getting, in order to catch issues like these when constructing tests. Change the styles of some panics.

CaspianA1 committed:
Fix some commented-out code

CaspianA1 committed:
Comment out a panic that was breaking the e2e tests

CaspianA1 committed:
Make some comments more descriptive

CaspianA1 committed:
Add a little TODO

qianl15 commented on dbos-rules.ts:
```diff
@@ -262,32 +255,198 @@ const awaitsOnNotAllowedType: DetChecker = (node, _fn, _isLocal) => {
-function analyzeFunction(fn: FunctionOrMethod) {
-  const body = fn.getBody();
+    // An injection in DBOS must match `ctxt.client.<something>`
+    if (identifiers.length !== 3) return;
```
Does this work if I assign `ctxt.client` to another variable and invoke `raw`?
```
const client = ctxt.client;
client.raw(`sql stmt`);
```

CaspianA1 replied:
Not yet - I have a TODO for that

chuck-dbos commented on dbos-rules.test.ts:
```diff
@@ -55,28 +55,244 @@ function makeExpectedDetCode(
+        // Literal concatenation (TODO: make adding parentheses in not fail)
+        ctxt.client.raw("foo" + "bar" + "baz" + "bam");
+
+        // Literal + variable concatenation
```
I would point out that this is OK because those variables have known values.

CaspianA1 replied:
Yes, it is - also, I fixed the parentheses thing, haven't pushed that yet

chuck-dbos commented:
Another thing on the future thought list... there are sanitization functions whose value can then be concatenated.  (Though usually easier just to use template, this comes up sometimes in DDLs.)

CaspianA1 committed:
So many changes (should have probably split up into more smaller commits). But here's a summary:
- Lots of little formatting/convention tweaks, and stuff like that
- Include some wrappers for getting symbols, to catch cases when a symbol isn't available
- Make `reduceNodeToLeftmostLeaf` more correct
- Skipping reporting expected-LR-usages of vars when they're used before they're declared (excluding `var` declarations)
- Now unpacking parenthesized expressions properly for LR-evaluation
- Come up with another way to represent the stack; the two ways have different benefits and drawbacks
- Checked diagnostics for all the tests - and I found so many problems there! Fixed them all, but it took a really long time.
- Report a possible error due to TSC API versioning when panicking for non-statemented nodes

CaspianA1 committed:
Make the e2e test a bit more general

CaspianA1 committed:
Upgrade the package number

CaspianA1 committed:
Follow semver for the new version

CaspianA1 committed:
I came up with a plan to fix most dependency issues. Part of that plan is to set peer dependencies and overrides here. The other one involves not installing typescript-eslint locally, and just installing it via this package. Aside from that, I made the versioning error message more descriptive, and removed an old TODO. I also upgraded the major version due to the versioning changes.

CaspianA1 committed:
Draft support for more callsites. Now supporting reassignment of `ctxt.client` to something else. Change some formatting. Change an error message name. Add back an expected test failure for `Date.now()`.

CaspianA1 committed:
Made a custom 'Maybe' type for all possibly undefined things

CaspianA1 committed:
Make the type-getting functionality clearer

CaspianA1 committed:
Now supporting Prisma as a callsite. Added some tests for that. The SQL injection checking now also checks all args for SQL injection. Make type getting work for literal types too now.

CaspianA1 committed:
Make some code more succinct

CaspianA1 committed:
Add a TODO for later

CaspianA1 committed:
Make line number reporting possible, finally! Also make some more variables const.

CaspianA1 committed:
Move a comment

CaspianA1 committed:
Add a semicolon that I forgot

CaspianA1 committed:
Now adding support for numeric literals

CaspianA1 committed:
Upgrade some docs + tests for numeric literals

CaspianA1 committed:
Some minor cleanup. Leave a comment.

CaspianA1 committed:
Add a blank line

CaspianA1 committed:
Make a kind check more idiomatic

CaspianA1 committed:
Fix some diagnostic errors

CaspianA1 committed:
Update some peer dependencies

CaspianA1 committed:
Make the module export strongly typed for the rule

CaspianA1 committed:
Slight reformatting. Add a TODO.

CaspianA1 committed:
Tweak package.json again

CaspianA1 committed:
Add a PoolClient test. Can also now match against many more cases for raw SQL clients.

CaspianA1 committed:
Last package.json version

CaspianA1 committed:
Do some deduplication

CaspianA1 committed:
Now supporting the remaining clients

CaspianA1 committed:
Merge 2 TODOs into 1

CaspianA1 committed:
Leave a big comment about the biggest bug at hand

CaspianA1 committed:
Now only checking the first argument again for raw SQL queries

CaspianA1 committed:
Final final version of dependencies! Will upgrade the docs to note additional installation instructions later.

CaspianA1 committed:
A bit of cleanup

CaspianA1 committed:
Add 2 more unit tests

CaspianA1 committed:
Remove peer dependencies

CaspianA1 committed:
Now checking all functions for SQL injection, not just transaction functions

CaspianA1 committed:
Add a useful link

CaspianA1 committed:
Add a future idea

CaspianA1 committed:
Change some var names

CaspianA1 committed:
Slight formatting tweak

CaspianA1 committed:
Make some naming more clear

CaspianA1 committed:
Remove a comma

CaspianA1 committed:
Now, if vars used for sql queries are not local, the whole program scope is examined.

kraftp commented on dbos-rules.ts:
```diff
@@ -348,28 +678,36 @@ function analyzeRootNode(eslintNode: EslintNode, eslintContext: EslintContext) {
     }
     else {
-      throw new Error(`Was expecting a statemented node! Got this kind instead: ${tsMorphNode.getKindName()}`);
+      const possibleVersioningError = `\
```
As discussed, we actually care about the version of the typescript compiler used by `typescript-eslint`, which constructs the AST. We should figure out a range of compatible versions (those using TS 5.4.x, for now) and fail on startup if the app uses any version of `typescript-eslint` out of that range.

CaspianA1 committed:
Clean up some old code

CaspianA1 committed:
Now failing with a more descriptive error message for versioning issues

CaspianA1 committed:
Tweak the versioning error

CaspianA1 committed:
Improve a variable name
